### ISSUE TITLE
[Definition and Roadmap] Referrals

### ISSUE THREAD

**Issue by clatwell** [Good-Bids/goodbids#363]
TL;DR - Registered Users can earn a Free Bid when a User who created an account on GoodBids via the original User's unique referral link places a Paid Bid on any Auction. 

### Requirements 

- Registered Users can access / copy a unique referral URL that they can share with friends 
- When a new User creates an account via the original User's referral URL, the system tracks the new User's association with the original User 
- When the new User places a Paid Bid on an Auction, the original User is awarded a Free Bid 

### Are there any existing plugins we can leverage? 

- The best option I could find was the [Referral System for WooCommerce](https://woo.com/products/referral-system-for-woocommerce/) plugin, which requires [WooCommerce Points and Rewards](https://woo.com/products/woocommerce-points-and-rewards/) as a dependency.  
- Ultimately I'm not sure this combo of plugins will do us any favors, because... 
  - It assumes the reward is granted as soon as the new User registers, whereas we need to grant the reward (a Free Bid) when the new User places their first Bid 
  - It assumes a "points" system, which doesn't match how we (currently) think about Free Bids 
  - There's a lot of extra functionality included in both of these plugins that we don't need 
- My initial hunch is that we could build our own functionality to support Referrals in less time than it would take us to extend / hack these plugins to do what we need, but that hunch needs validation from Brian. 

### Technical Approach 

Assuming we build this without using an existing plugin, I'm imagining the work would look something like this... 

**See it Work** 

- Create a `Referrals` join table at the network level that stores the id of the sender and receiver user -- David [blogged about this](https://www.viget.com/articles/friends-undirected-graph-connections-in-rails/) a few years ago, might be some good tips there -- and maybe also store a boolean for "hasConverted?"   
- Add the ability to append a User-specific query string to the `/my-account` URL that would enable us to tie a new registration to an existing User 
- Display a button on the Auction page below the Auction title ("Share for free bids") that copies the User's unique Referral URL to their clipboard 
- When a new User creates an account via a query stringed URL, create an entry in the Referrals table with the id of the sender and receiver 
- When a User places a bid, check the Referrals table to see if they were the receiver of a Referral and if `hasConverted?` is false. If yes to both, flip `hasConverted?` to true and award a Free Bid to the sender. 
  - This will probably require some refactoring to Free Bids so we can support FBs earned from bidding and from referring. 
- Display the referral-earned Free Bid on the Free Bid page in the My Account section for the sender... it might be easiest to display these in a separate table from the Free Bids earned from bidding 

**Make it Better** 

- Add each User's unique Referral URL in the GET User endpoint so the Bidding Block can display it and Users can access it directly from the Auction page
- Create a "Referrals" page in the My Account section that shows a table of Referrals created and converted and allows a User to copy their unique Referral URL to their clipboard 
- Add a UI component to the logged-out view of the `/my-account` page that conditionally displays if the User has a query stringed URL for the page that says, "You've been referred by `userName`" or something similar. 
- When a new User creates an account from a referral link, redirect them to the All Auctions page associated with the Nonprofit from whence the referral link was generated 
- Create a second custom block for the Auction page that displays the User's custom referral link and allows them to copy to their clipboard / post to social media 
- When a user clicks the "Share for free bids" CTA on the Auction page, open a modal / overlay with copy to clipboard / post to social media options 

### Open Questions 

- The only flag I can see here is that the Referral URL will be both User and Nonprofit-site specific, as there's (currently) no generic logged-out `/my-account` view with the registration form at the GB main site level. 
- I think this is fine, and will probably make the redirect-after-registration enhancement easier. 
- But, if we needed to, we could add support for registration at the GB main site level (although this could get tricky since the GB main site is technically not a WC store with customers)
